### PR TITLE
[plugins] return err when failing remove

### PR DESCRIPTION
**Pull request by tophj-ibm**
Fixes a case where removing the plugin from disk would
fail silently. Also moves pluginStore remove after we
remove from disk, so 'docker plugin ls'  isn't
empty in case it errors out.

Signed-off-by: Christopher Jones <tophj@linux.vnet.ibm.com>

ping @anusha-ragunathan @tiborvass

### CODE REVIEW

**Comment by tophj-ibm** [comment]
I've been trying to write a test case for this, but I can't figure out how to get RemoveAll to fail when run as root. Even the go stdlib RemoveAll test fails in this case.

**Comment by tiborvass** [comment]
Seems reasonable. LGTM

File: plugin/backend_linux.go [hunk 1]
```diff
@@ -294,11 +294,11 @@ func (pm *Manager) Remove(name string, config *types.PluginRmConfig) error {
 	}
 
 	id := p.GetID()
-	pm.pluginStore.Remove(p)
 	pluginDir := filepath.Join(pm.libRoot, id)
 	if err := os.RemoveAll(pluginDir); err != nil {
-		logrus.Warnf("unable to remove %q from plugin remove: %v", pluginDir, err)
+		return fmt.Errorf("unable to remove %q from plugin remove: %v", pluginDir, err)
```
Comment (thaJeztah) [inline plugin/backend_linux.go]: was discussing with @tiborvass; should this use `errors.Wrap()`?

Comment (thaJeztah) [inline plugin/backend_linux.go]: It's provided through this package; https://github.com/pkg/errors#adding-context-to-an-error

Comment (tophj-ibm) [inline plugin/backend_linux.go]: thanks, I did find and read that 😄  PR should be up to date now.

File: plugin/backend_linux.go [hunk 2]
```diff
@@ -294,11 +294,11 @@ func (pm *Manager) Remove(name string, config *types.PluginRmConfig) error {
 	}
 
 	id := p.GetID()
-	pm.pluginStore.Remove(p)
 	pluginDir := filepath.Join(pm.libRoot, id)
 	if err := os.RemoveAll(pluginDir); err != nil {
-		logrus.Warnf("unable to remove %q from plugin remove: %v", pluginDir, err)
+		return errors.Wrap(err, "failed to remove plugin directory")
```
Comment (cpuguy83) [inline plugin/backend_linux.go]: Probably need to handle this like in container rm where it's something like
```go
defer func() {
    if err == nil || config.ForceRemove {
        pm.pluginStore.Remove(p)
        pm.pluginEventLogger(id, name, "remove")
    }
}()

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

File: plugin/backend_linux.go (hunk 2 above)
Comment (tophj-ibm) [inline plugin/backend_linux.go]: question: If we are force removing and RemoveAll fails, shouldn't it keep the plugin in the store so you can see that it is still there / that remove failed (besides the err)? If it's not in the store, doesn't that means the only way to remove it is by manually removing it from disk?

Comment (cpuguy83) [inline plugin/backend_linux.go]: I think if someone is specifying force remove then we should remove it from the store no matter what.
We could report a warning still.

This is what we do for containers and volumes.

Comment (tophj-ibm) [inline plugin/backend_linux.go]: Okay, should be updated. It will throw an error on that case anyway.

File: plugin/backend_linux.go [hunk 3]
```diff
@@ -294,12 +294,18 @@ func (pm *Manager) Remove(name string, config *types.PluginRmConfig) error {
 	}
 
 	id := p.GetID()
-	pm.pluginStore.Remove(p)
+
+	defer func() {
+		if err == nil || config.ForceRemove {
```
Comment (cpuguy83) [inline plugin/backend_linux.go]: you'll need to name the error return for this to work, or change `if err :=` to `if err = `. and define error above.

**Comment by vieux** [comment]
ping @cpuguy83

**Review by cpuguy83 — APPROVED** [review]
LGTM

**Review by vdemeester — APPROVED** [review]
LGTM 🐻

**Pull request merged**